accepted/00000-hub-reporting.md

```diff
@@ -0,0 +1,177 @@
+In summary:
+- DB in Uyuni Hub stores, collects and eventually aggregates data coming from all the DB in Uyuni Server,
+- DB in Uyuni Server stores it own data,
+- The reporting tool can be connected either to the Hub or to any Uyuni Server,
+- all connections are secure.
+
+### The Database Schema
+
+The schema should export the most important tables from the main Uyuni Database as a slightly de-normalized
+variant containing only data which are relevant for a report.
+
+A ready-to-use report can be provided with views doing joins over multiple tables.
+No foreign keys should be used to make data update and refreshing easier and independent from the order
+of the tables.
+
+Every table gets an extra column for the Uyuni server id of server which provide the data (`mgm_id`).
+On a single Uyuni Server this column has the standard value `1` which represent "localhost".
+On the Hub it will be replaced with the real server id the managed server has in the hub database.
+
+Every row should get a report timestamp column (`exported`), which is set to the time when the data
+were exported from the main Uyuni Server database.
+
+Indexes should be set on the typical columns which might be searched for.
+
+Data which belong to the traditional stack only (like osad status) should not be made available in the
+reporting database as the traditional stack will be deprecated soon.
+
+A detailed Database schema is not part of this RFC.
+
+# Installation
+The reporting tools will be installed by default; all the upgrade to a newer version, will install the 
+reporting tools as well. Several packages will be provided in order to provide the following tools and
+functionalities:
+- setup the database and initialize the schema,
+- perform database schema migration,
+- setup database users,
+- setup tasktomatic jobs.
+
+All the other information required (e.g. Uyuni Server info) will be retrieved by already existing configuration.
+
+## Reporting Database
+A reporting DB is setup in all Uyuni Servers and Uyuni Hub nodes to store reporting data.
+The default setup will create a Read/Write user, to manage the database and write the data from it into the Reporting DB.
+This user and the DB connection parameters are written into `/etc/rhn/rhn.conf` similar to the default DB options.
+
+## Initial Setup
+Uyuni Hub should create its own read-only user on the Reporting Databases of the single Servers.
+As the Servers are managed with salt, we will write a state to create an account.
+The username and password are generated on the Hub and provided as pillar data.
+On the server the state take care of the existence of the account and the Hub
+can store the parameters in its database under the system entry.
+
+# Workflow
+
+## Scheduled job: collect from Uyuni DB
+A taskomatic job is responsible to fetch and prepare the data from the main Uyuni
+Database and insert them into the Reporting Database.
```

Comment:
What pros and cons do you see in using [Postgres' materialized views](https://www.postgresql.org/docs/13/rules-materializedviews.html) instead of Java code to do the selectiong/processing/inserting?

The Taskomatic job would then have to call `REFRESH MATERIALIZED VIEW reportingschema.reportingtable` for each table in the reporting schema.

Reply:
We design already for using an external DB for the report. Materialized Views do not work with a different database and not over network when it is also external.

Reply:
OTOH, all denormalizing code would be in Java, which among other things means that columns have to be mapped and when the schema changing mapping code will also need to be adapted accordingly. Moreover Java mapping/unmapping could become a funnel performance-wise.

With views, we will know when the view breaks because of an underlying schema change, as it will not compile. Also, I see that that some trivial changes like adding a column that was not there before could be easier than changing the appropriate mapping on the appropriate classes with the appropriate types.

I see the objection of contemplating the external DB as a deployment possibility from the start and I agree it is a valid one. Still, I feel like the approaches are not mutually exclusive - we could for example still keep a (non-materialized) view on the peripheral server's schema, then copy over data from it to a (potentially external DB) plain table. That copy could be implemented "blindly" ie. just assuming that the schemas match, `SELECT` from one side and `INSERT` into the other - with no object mapping at all.

How does that sound?

Reply:
We never had the idea to write a real hibernate object mapping. We suggested to use mode queries with select and insert.
Using a view and do a blind "select * from" and "insert into VALUES()" is a good idea in my opinion.

Reply:
JFYI - very very similar automated-schema-inspection and generic (safe) SELECT/INSERT code exists in the ISSv2 codebase - in Golang of course.

With this comment I am not necessarily advocating to have a self-contained Golang tool based on that code to perform the data copy - I am just mentioning it as an alternative option for your consideration.